### ISSUE TITLE
Accesses to module fields are compiled as mutable loads

### ISSUE THREAD

**Issue by OlivierNicole** [ocaml/ocaml#11654]
I realized that accesses to fields of a module are treated as mutable loads, whereas they shoud be considered immutable. This could cause a performance regression between 4.14 and 5.x on architectures where mutable loads are compiled differently from immutable loads, such as ARM.

The access path is first correctly translated to a sequence of immutable reads in Lambda
https://github.com/ocaml/ocaml/blob/bd87ae0632d352a9a15869c6047888c319130fd3/lambda/lambda.ml#L666-L673
but that `Immutable` attribute is then ignored when translating to C−−:
https://github.com/ocaml/ocaml/blob/bd87ae0632d352a9a15869c6047888c319130fd3/asmcomp/cmmgen.ml#L801-L802

**Comment by xavierleroy** [comment]
>  architectures where mutable loads are compiled differently from immutable loads, such as ARM

It seems to me that nonatomic loads from mutable locations are always compiled as plain "load" instructions, even on ARM64.  It's the atomic loads that need special treatment on ARM64.

At any rate, looking at the ARM64 assembly generated by ocamlopt on trunk for module field accesses, I don't see any barriers.

**Comment by kayceesrk** [comment]
@xavierleroy’s observation is correct. It is only the non-atomic stores that get a fence on ARM64, not non-atomic loads. The fence before a non-atomic store ensures that prior loads are not reordered after the store, preserving the load-store ordering.

**Comment by lpw25** [comment]
IIRC the notion of immutable read in cmm and the one in lambda/flambda/clambda are not quite the same, and that is why the immutability is currently dropped in cmmgen. In particular, due to the compilation of recursive values and recursive modules (since #10205), the reads marked immutable in lambda need to be careful not to be moved before some kinds of writes, and I don't think that selectgen would ensure that if these reads were marked immutable in cmm.

**Comment by xavierleroy** [comment]
Yes, there's something fishy with recursive modules (as usual), and it's quite likely that we miss some opportunities for CSE of loads corresponding to module accesses because of that.  But I think this problem is already present in OCaml 4 and is not made any worse by OCaml 5.

**Comment by OlivierNicole** [comment]
I see, so it is not a regression, I stand corrected.

In addition to missing CSE opportunities, I had to change this for our work on ThreadSanitizer instrumentation (where we really want to avoid instrumenting memory reads if we can). So far, we did not encounter bugs due to that. But by what @lpw25 writes, maybe we will…

**Comment by stedolan** [comment]
To elaborate slightly: Here are three definitions of "immutable", from strongest to weakest:

  1. There is only one write, initialisation.
  2. All writes precede all reads, so all reads return the same value.
  3. Reads and writes mix arbitrarily, but all reads return semantically equivalent values.

Case (1) covers most immutable fields in OCaml. Case (2) covers `let rec` on values and TMC: there are multiple writes (first a dummy and then a final value), but `Rec_check` ensures that the dummy is never read. Case (3) covers `module rec`: user code and module initialisation are mixed, but enough backpatching is done that the old and new contents of a field are not semantically distinguishable (except by physical equality).

Cases (1) and (2) are enough to justify CSE, and case (3) also justifies CSE if you're happy for it to possibly affect physical equality of functions and lazy values (which is semantically not fully defined anyway).

However, only case (1) is enough to justify the transformation @lpw25 mentions above, where an immutable read is moved before an arbitrary non-initialising write. In case (3), this is permissible again up to changing physical equality, but in case (2) it's broken if the write and the read alias, because a dummy value could become visible.

In short, I think CSE is fine on module blocks or anywhere else (it may affect physical equality of recursive module contents, but so what), but other transformations are not valid. (Sadly, I'm not sure that Selectgen currently makes a fine enough distinction between the ones that are and aren't valid here).

**Comment by xavierleroy** [comment]
That was an interesting discussion.  Thanks to @stedolan for the summary of the many notions of immutability.  Maybe we're losing a few opportunities for CSE here (?).  On the other hand, the code generated to access components of modules is very decent already.  So, I'm going to close this issue, while encouraging anyone who feels strongly about immutability-based back-end optimizations to start a discussion or propose a PR.
